File: ICPC/Codecamp 2023 DUT_ICPC/D_Best_Wishes.py
MAX = 1000000  # Maximum value of D can be 10^6

t = int(input())

dp = [float('inf')] * (MAX + 1)  # initialise dp[i] with infinity
# Here dp[i] = min. number of days to make the charges i on a single day.

dp[1] = 1  # Initial charge is Rs. 1, so no extra days required

for i in range(2, MAX + 1):
    if i % 2 == 0:
        dp[i] = min(dp[i], dp[i // 2] + 1)
    if i % 3 == 0:
        dp[i] = min(dp[i], dp[i // 3] + 1)
    if i - 1 >= 1:
        dp[i] = min(dp[i], dp[i - 1] + 1)

for _ in range(t):
    n = int(input())

    # minimum number of days to make the charges for a single day, equal to n
    print(dp[n])

    seq = []

    # Now, perform a traceback to get the sequence of charges on each day.
    i = n
    while i > 1:
        seq.append(i)
        if i - 1 >= 1 and dp[i] == 1 + dp[i - 1]:
            i = i - 1
        elif i % 2 == 0 and dp[i] == 1 + dp[i // 2]:
            i = i // 2
        elif i % 3 == 0 and dp[i] == 1 + dp[i // 3]:
            i = i // 3

    seq.append(1)

    seq.reverse()

    print(' '.join(map(str, seq)))

File: ICPC/Codecamp 2023 DUT_ICPC/D_Best_Wishes.cpp
#include <iostream>
#include <vector>
#include <limits>
#include <algorithm>

const int MAX = 1000000; // Maximum value of D can be 10^6

int main()
{
    int t;
    std::cin >> t;

    std::vector<int> dp(MAX + 1, std::numeric_limits<int>::max());

    dp[1] = 1;

    for (int i = 2; i <= MAX; i++)
    {
        if (i % 2 == 0)
        {
            dp[i] = std::min(dp[i], dp[i / 2] + 1);
        }
        if (i % 3 == 0)
        {
            dp[i] = std::min(dp[i], dp[i / 3] + 1);
        }
        if (i - 1 >= 1)
        {
            dp[i] = std::min(dp[i], dp[i - 1] + 1);
        }
    }

    while (t--)
    {
        int n;
        std::cin >> n;

        std::cout << dp[n] << '\n';

        std::vector<int> seq;

        int i = n;
        while (i > 1)
        {
            seq.push_back(i);
            if (i - 1 >= 1 && dp[i] == 1 + dp[i - 1])
            {
                i = i - 1;
            }
            else if (i % 2 == 0 && dp[i] == 1 + dp[i / 2])
            {
                i = i / 2;
            }
            else if (i % 3 == 0 && dp[i] == 1 + dp[i / 3])
            {
                i = i / 3;
            }
        }

        seq.push_back(1);

        std::reverse(seq.begin(), seq.end());

        for (int i = 0; i < seq.size(); i++)
        {
            std::cout << seq[i] << ' ';
        }
        std::cout << '\n';
    }

    return 0;
}
